Doc: Update BT Mesh docs after extracting regulator (#6818)
nrfconnect/sdk-nrf: doc/nrf/libraries/bluetooth_services/mesh/light_ctrl_reg.rst

.. _bt_mesh_light_ctrl_reg_readme:

Illuminance regulator interface
###############################

This module defines an interface for implementation and interaction with an abstract illuminance regulator.

Using the regulator
*******************

As this defines an interface for an abstract illuminance regulator, it cannot be used on its own.
For an example of a regulator usage, see :ref:`bt_mesh_light_ctrl_srv_readme`.

A regulator implementation will provide a :c:struct:`bt_mesh_light_ctrl_reg` instance, and this is used as the interface to the regulator.
Field :c:member:`bt_mesh_light_ctrl_reg.user_data` of this struct can be used to store a user context.

> Before using the regulator, it needs to be initialized through :c:member:`bt_mesh_light_ctrl_reg.init` function.

Reviewer: ```suggestion
Before using the regulator, it needs to be initialized through the :c:member:`bt_mesh_light_ctrl_reg.init` function.
```

Functions :c:member:`bt_mesh_light_ctrl_reg.start` and :c:member:`bt_mesh_light_ctrl_reg.stop` should be used to start and stop the regulator.

Regulator inputs
----------------

The regulator has two inputs:
* target value
* measured value

The illuminance regulator takes the specified target value as the reference level and compares it to the reported measured value.
The inputs are compared to establish an error for the regulator, which the regulator tries to minimize.

The measured value should be passed through :c:member:`bt_mesh_light_ctrl_reg.measured` variable.
The passed value is to be used in the next regulator step.
The regulator depends on a frequency of the measurements to provide a stable output.

> If the measurement frequency is too low, the regulator might oscillate, as it attempts to compensate for outdated feedback.

Author: I added this phrase in between:
```
The measurement frequency should be as close as possible to the update interval of the regulator.
```

The value of the coefficients is typically a trade-off between fast response time and system instability:

* If the value is too high, the system might become unstable, potentially leading to oscillation and loss of control.
* If the value is too low, the step response might be too slow or unable to reach the target value altogether.

Implementing a new regulator
****************************

To implement a new regulator using this generic interface, declare a :c:struct:`bt_mesh_light_ctrl_reg` struct, and set the :c:member:`bt_mesh_light_ctrl_reg.init`, :c:member:`bt_mesh_light_ctrl_reg.start`, and :c:member:`bt_mesh_light_ctrl_reg.stop` fields to implementations of these functions.
Use :c:func:`bt_mesh_light_ctrl_reg_input_get` to get the current target value for the regulator supplied by the regulator user.
The value returned is interpolated linearly over the transition time, if a transition time is requested by the regulator user.

The :c:member:`bt_mesh_light_ctrl_reg.init` function should perform the necessary steps to initialize the implementation, such as initializing interrupt handlers or timers, but it will not start the regulator.

> The :c:member:`bt_mesh_light_ctrl_reg.start`, and :c:member:`bt_mesh_light_ctrl_reg.stop` functions should start and stop the regulator, after it has been initialized by a call to ``init``.
> 
> On every regulator step, the regulator should call :c:member:`bt_mesh_light_ctrl_reg.updated` callback supplied by the user.
> 
> For an example of a regulator implementation, see :ref:`bt_mesh_light_ctrl_reg_spec_readme`.

Reviewer: ```suggestion
Use the :c:member:`bt_mesh_light_ctrl_reg.start` and :c:member:`bt_mesh_light_ctrl_reg.stop` functions to start and stop the regulator after it has been initialized by a call to ``init``.

On every regulator step, the regulator must call :c:member:`bt_mesh_light_ctrl_reg.updated` callback supplied by the user.

For an example of regulator implementation, see :ref:`bt_mesh_light_ctrl_reg_spec_readme`.
```

Light Lightness Control regulator
#################################

This module is an abstract proportional-integral (PI) regulator.
Regulator implementations must declare a :c:struct:`bt_mesh_light_ctrl_reg` struct and supply implementations for ``init``, ``start``, and ``stop`` functions.

For an example of a regulator implementation, see :ref:`bt_mesh_light_ctrl_reg_spec_readme`.

Using the regulator
*******************

> As this implements an abstract regulator, it cannot be used on its own.

Reviewer: How about: As this is an abstract regulator, it cannot be used on its own.

For an example of a regulator implementation, see :ref:`bt_mesh_light_ctrl_reg_spec_readme`.

Using the regulator
*******************

> As this module implements an abstract regulator, it cannot be used on its own.

Reviewer 2: ```suggestion
As this implements an abstract regulator, it cannot be used on its own.
```

A regulator implementation will provide a :c:struct:`bt_mesh_light_ctrl_reg` instance, and this is used as the interface to the regulator.
Users of a regulator will call ``init``, ``start``, and ``stop`` functions from this instance, as well as :c:func:`bt_mesh_light_ctrl_reg_input_set`.

> To get output from the regulator, set the ``update`` callback on the struct.

Reviewer 2: ```suggestion
To get output from the regulator, set the ``update`` callback in the struct.
```

There is a field in the struct for ``user_data`` which can be supplied by the user and will be available through the regulator pointer passed to the ``update`` callback.

Implementing a regulator
************************

To implement a regulator using this generic interface, declare a :c:struct:`bt_mesh_light_ctrl_reg` struct, and set the ``init``, ``start``, and ``stop`` fields to implementations of these functions.
To get the current target value for the regulator, supplied by the user of the regulator, use :c:func:`bt_mesh_light_ctrl_reg_input_get`, which returns the requested target value.
The value returned is interpolated linearly over the transition time, if a transition time is requested by the regulator user.

> The ``init`` should perform any necesary steps to initialize the implementation, such as initializing interrupt handlers or timers, but not start the regulator.

Reviewer 2: ```suggestion
The ``init`` should perform any necessary steps to initialize the implementation, such as initializing interrupt handlers or timers, but not start the regulator.
```

> The ``start`` and ``stop`` functions should start and stop the regulator, after it has been initialized by a call to ``init``.

Reviewer 2: ```suggestion
The ``start`` and ``stop`` functions will start and stop the regulator, after it has been initialized by a call to ``init``.
```

> When the regulator is running, it should repeatedly call the ``update`` function on the :c:struct:`bt_mesh_light_ctrl_reg` struct, which is supplied by the user.

Reviewer 2: ```suggestion
When the regulator is running, it will repeatedly call the ``update`` function on the :c:struct:`bt_mesh_light_ctrl_reg` struct, which is supplied by the user.
```
